### DISCUSSION TITLE
Core AAC encoder and sampling rate issue

### DISCUSSION THREAD

**Discussion by AlanHK** [enzo1982/freac#448] [Audio Format Discussion]
I often make Audiobook M4B files from sets of MP3 files.
I use the Core AAC encoder, from Windows iTunes 12.1.

Recently I found one set of MP3 files that every time I tried this, freac would proceed until it had processed about 2 hours of files, then abort. Most books are over 8 hours. 
It wasn't a bad file. If I put them in a different order, it would crash on a different files, but at about the same total duration.
If I converted to MP3s at a different bitrate, that had no problem.

I have made a few dozen audiobook M4Bs now, and never had this issue.
I checked the media info and eventually noticed that the MP3s had a sampling rate of 16kHz, which is unusual. But plays fine everywhere.

```
Format : MPEG Audio 
Format version : Version 2
Format profile : Layer 3 
Duration : 30s 348ms
Bit rate mode : Constant
Bit rate : 32.0 Kbps
Channel(s) : 2 channels
Sampling rate : 16.0 KHz
Compression mode : Lossy
```
 
So in the freac config I added I added the Sample Rate Converter and set it to change the rate to 22.05 kHz.
Now the creation of the M4B completes. Perhaps a bit slower, but not a problem.

I then changed the sample rate setting to 16kHz. I loaded a set of MP3s that had previously converted to M4B and now it also crashed after processing a couple of hours.

So apparently Core, at least the version I have, has a problem with 16kHz sampling. 
Though it can convert short 16kHz sets, it crashes on longer ones.

**Comment by enzo1982** [comment]
Hi Alan, you already reported this in an email. Didn't you get my response?

I can reproduce the issue here using the Core Audio Codec files from iTunes 12.1. This appears to be caused by a memory leak in the High Efficiency profile encoder on Apple's side, so unfortunately there is nothing I can do about it. The memory leak causes the codec to use more and more memory during encoding and at some point it will exceed either the amount of available RAM or the 32 bit limit of 2 GB per process and crash the application. When using a newer version of the Core Audio Codec, the issue goes away, but unfortunately that's no option on XP, as there is no newer version for it.

Using the Sample Rate Converter in fre:ac to change to a different sample rate before feeding the samples to the encoder should be a viable work-around. Alternatively, you could use the Low Complexity profile instead of the High Efficiency one. For bit rates above 64 kbps total (so 32 kbps per channel for Stereo), the Low Complexity profile is recommended for best quality anyway. As another alternative, you should be able to use the FDK-AAC encoder instead of the Core Audio one.

**Comment by AlanHK** [comment]
Thanks, your email went to spam and I have found it now. 

I use a low bitrate for most audiobooks, which are single channel, one narrator, no music. The profile I use is 24kbps per channel. (Would like to have a single channel option.)
Slightly higher rate for dramatised readings. 
So I'll keep High Efficiency and change the sample rate.

**Reply by enzo1982** [reply]
> (Would like to have a single channel option.)

To create single channel M4Bs from Stereo MP3s, just use the Channel Converter (in the _Processing_ settings) set to Mono.

**Comment by AlanHK** [comment]
I do that, but the resulting M4B is still stereo, according to MediaInfo.
I think single channel AAC is allowed, but there is no way to set it.

Info for the M4B file discussed above:

```
ID : 1  Format : AAC 
Format : AAC  Format
Format/Info : Advanced Audio Codec
Format profile : HE-AAC / LC 
Codec ID : 40  Duration : 11h 5mn 
Bit rate mode : Variable 
Bit rate : 23.8 Kbps 
Maximum bit rate : 37.6 Kbps 
Channel(s) : 2 channels 
Channel(s)_Original : 1 channel 
Channel positions : Front: C 
Sampling rate : 32.0 KHz / 16.0 KHz 

```

But I also see here that the sampling rate is reported as  16.0 KHz , when as above it was converted to 22.05. So maybe the media info isn't accurate?

I guess AAC uses something like joint stereo in MP3, so there may not be much overhead with that and actual mono when both channels are identical.

**Reply by ValeZAA** [reply]
Mono HE-AAC is always treated as stereo! Because it can become stereo at any point, spec permits that. All decoders in ffmpeg, aac_fixed, aac float, aac_latm and libfdk_aac treat mono as stereo, ALWAYS.
